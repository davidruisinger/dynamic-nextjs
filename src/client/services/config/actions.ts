import { createAction } from 'typesafe-actions'
import { Config } from './types'

/**
 * Redux action to set the shopId.
 * @param {Object} config
 * @param {string} config.shopId - The shopId for this client.
 */
export const setShopId = createAction('SET_SHOP_ID', resolve => {
  return ({ shopId }: { shopId: string }) => resolve({ shopId })
})

/**
 * Redux action to set the baseUrl for config requests.
 * @param {Object} config
 * @param {string} config.baseUrl - The baseUrl of the config service.
 */
export const setBaseUrl = createAction('SET_CONFIG_BASE', resolve => {
  return ({ baseUrl }: { baseUrl: string }) => resolve({ baseUrl })
})

/**
 * Redux action to fetch the config from remote.
 */
export const fetchConfig = createAction('GET_CONFIG_REQUEST', resolve => {
  return () => resolve()
})

/**
 * Redux action to process the fetched config.
 * @param {Object} result
 * @param {string} result.config - The fetched config object.
 */
export const fetchConfigSuccess = createAction(
  'GET_CONFIG_SUCCESS',
  resolve => {
    return ({ config }: { config: Config }) => resolve({ config })
  }
)

/**
 * Redux action to process an error while fetching the config.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchConfigFail = createAction('GET_CONFIG_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
