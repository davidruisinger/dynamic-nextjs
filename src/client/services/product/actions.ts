import { createAction } from 'typesafe-actions'
import { Product } from './types'

/**
 * Redux action to fetch multiple products by ID.
 * @param {Object} config
 * @param {string[]} config.productIds - The productIds to fetch.
 */
export const fetchProducts = createAction('GET_PRODUCTS_REQUEST', resolve => {
  return ({ productIds }: { productIds: string[] }) => resolve({ productIds })
})

/**
 * Redux action to process an array of fetched products.
 * @param {Object} result
 * @param {string} result.products - The fetched products.
 */
export const fetchProductsSuccess = createAction(
  'GET_PRODUCTS_SUCCESS',
  resolve => {
    return ({ products }: { products: Product[] }) => resolve({ products })
  }
)

/**
 * Redux action to process an error while fetching products.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchProductsFail = createAction('GET_PRODUCTS_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
