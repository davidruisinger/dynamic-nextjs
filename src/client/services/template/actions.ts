import { createAction } from 'typesafe-actions'
import { Template } from './types'

/**
 * Redux action to fetch a template from remote.
 * @param {Object} config
 * @param {string} config.shopId - The shopId for which the template should be fetched.
 */
export const fetchTemplate = createAction('GET_TEMPLATE_REQUEST', resolve => {
  return () => resolve()
})

/**
 * Redux action to process a fetched template.
 * @param {Object} result
 * @param {string} result.template - The fetched template object.
 */
export const fetchTemplateSuccess = createAction(
  'GET_TEMPLATE_SUCCESS',
  resolve => {
    return ({ template }: { template: Template }) => resolve({ template })
  }
)

/**
 * Redux action to process an error while fetching a template.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchTemplateFail = createAction('GET_TEMPLATE_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
