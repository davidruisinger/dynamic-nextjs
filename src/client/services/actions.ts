// We do not really need this except for generating the global Action

import * as configActions from './config/actions'
import * as templateActions from './template/actions'
import * as productActions from './product/actions'

export default {
  config: configActions,
  template: templateActions,
  product: productActions,
}
