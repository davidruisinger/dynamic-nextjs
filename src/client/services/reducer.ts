import { combineReducers } from 'redux'

// Import each service reducer and it's name
import { SERVICE_NAME as configServiceName } from './config/constants'
import ConfigReducer from './config/reducer'
import { SERVICE_NAME as templateServiceName } from './template/constants'
import TemplateReducer from './template/reducer'
import { SERVICE_NAME as productServiceName } from './product/constants'
import ProductReducer from './product/reducer'

const rootReducer = combineReducers({
  [configServiceName]: ConfigReducer,
  [templateServiceName]: TemplateReducer,
  [productServiceName]: ProductReducer,
})

export default rootReducer
