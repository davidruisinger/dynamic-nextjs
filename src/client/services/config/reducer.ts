import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { Config } from './types'

export type ConfigState = Readonly<{
  shopId: string
  baseUrl: string
  config: Config
  isFetching: boolean
  error: string
}>

const initialState: ConfigState = {
  shopId: null,
  baseUrl: null,
  config: null,
  isFetching: false,
  error: null,
}

export default (
  state: ConfigState = initialState,
  action: ActionType<typeof actions>
): ConfigState => {
  switch (action.type) {
    case getType(actions.setShopId):
      return {
        ...state,
        shopId: action.payload.shopId,
      }

    case getType(actions.setBaseUrl):
      return {
        ...state,
        baseUrl: action.payload.baseUrl,
      }

    case getType(actions.fetchConfig):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchConfigSuccess):
      return {
        ...state,
        config: action.payload.config,
        isFetching: false,
      }

    case getType(actions.fetchConfigFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
