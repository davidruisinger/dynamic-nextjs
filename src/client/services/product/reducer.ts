import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { Product } from './types'

export type ProductState = Readonly<{
  products: Product[]
  isFetching: boolean
  error: string
}>

const initialState: ProductState = {
  products: [],
  isFetching: false,
  error: null,
}

export default (
  state: ProductState = initialState,
  action: ActionType<typeof actions>
): ProductState => {
  switch (action.type) {
    case getType(actions.fetchProducts):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchProductsSuccess):
      return {
        ...state,
        products: [...state.products, ...action.payload.products],
        isFetching: false,
      }

    case getType(actions.fetchProductsFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
