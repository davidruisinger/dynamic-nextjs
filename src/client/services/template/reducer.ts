import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { Template } from './types'

export type TemplateState = Readonly<{
  template: Template
  isFetching: boolean
  error: string
}>

const initialState: TemplateState = {
  template: null,
  isFetching: false,
  error: null,
}

export default (
  state: TemplateState = initialState,
  action: ActionType<typeof actions>
): TemplateState => {
  switch (action.type) {
    case getType(actions.fetchTemplate):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchTemplateSuccess):
      return {
        ...state,
        template: action.payload.template,
        isFetching: false,
      }

    case getType(actions.fetchTemplateFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
