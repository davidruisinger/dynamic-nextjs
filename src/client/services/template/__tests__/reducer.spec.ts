import reducer from '../reducer'
import * as actions from '../actions'
import { templateMock } from './_mocks'

const initialState = reducer(undefined, {} as any)

describe('template service reducer', () => {
  describe('initial state', () => {
    it('should match the snapshot', () => {
      expect(initialState).toMatchSnapshot()
    })
  })

  let fetchingState, successState, errorState
  describe('fetchTemplate action', () => {
    it('should set isFetching', () => {
      const action = actions.fetchTemplate()
      fetchingState = reducer(initialState, action)
      expect(fetchingState.isFetching).toBeTruthy()
    })
  })

  describe('fetchTemplateSuccess action', () => {
    it('should unset isFetching and set the template', () => {
      const action = actions.fetchTemplateSuccess({ template: templateMock })
      successState = reducer(fetchingState, action)
      expect(successState.isFetching).toBeFalsy()
      expect(successState.template).toEqual(templateMock)
    })
  })

  describe('fetchTemplateFail action', () => {
    it('should unset isFetching and set an error', () => {
      const errorMessage = 'Test error'
      const action = actions.fetchTemplateFail({ message: 'Test error' })
      errorState = reducer(fetchingState, action)
      expect(errorState.isFetching).toBeFalsy()
      expect(errorState.error).toBe(errorMessage)
    })
  })

  describe('fetchTemplate action', () => {
    it('should set isFetching and clear the error when re-trying', () => {
      const action = actions.fetchTemplate()
      const retryState = reducer(errorState, action)
      expect(retryState.isFetching).toBeTruthy()
      expect(retryState.error).toBe(null)
    })
  })
})
