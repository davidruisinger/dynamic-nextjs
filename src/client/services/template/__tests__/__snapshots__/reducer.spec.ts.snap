// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template service reducer initial state should match the snapshot 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "template": null,
}
`;
